File: perf/benchmarks/micro/closure_sum/closure_sum.go
// Closure / higher-order pipeline — micro-benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	n := int64(2_000_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		n = v
	}
	data := make([]int64, n)
	for i := int64(0); i < n; i++ {
		data[i] = i
	}

	// Use func-value variables (instead of inlinable literals) so the call is
	// dispatched via a closure pointer, matching the C++ std::function path.
	mapper := func(x int64) int64 { return x*x - 7 }
	pred := func(x int64) bool { return x%3 == 0 }
	reducer := func(a, b int64) int64 { return a + b }

	t0 := time.Now()
	var cs int64
	for i := int64(0); i < n; i++ {
		v := mapper(data[i])
		if pred(v) {
			cs = reducer(cs, v)
		}
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/micro/closure_sum/closure_sum.py
# Closure / higher-order pipeline — micro-benchmark.
#
# Builds an array of N integers, then evaluates a map -> filter -> reduce
# pipeline using closures. Stresses higher-order functions, closure dispatch,
# and integer arithmetic.
#
# Argument: N (default 2_000_000).

import sys
import time
from functools import reduce


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 2_000_000
    data = list(range(n))

    t0 = time.perf_counter_ns()
    cs = reduce(
        lambda acc, x: acc + x,
        filter(lambda x: x % 3 == 0, map(lambda x: x * x - 7, data)),
        0,
    )
    elapsed_ns = time.perf_counter_ns() - t0

    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
